# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [161]:
import pandas as pd
from src.utils import load_data_from_google_drive

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [162]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [163]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [164]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [165]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [166]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [167]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [168]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [169]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [170]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [171]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [172]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [173]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

## **Q1**: How many unique seasons are there in the dataset? Use the nunique() function.

When calling this function, we extract the amount of unique/non-duplicate values in the the dataset for each attribute. When we specifically want to know how many unique seasons there are, we specify which attribute we're interested in in the bracketed parentheses. I have shown how to do so below. 

In [174]:
df['Season'].nunique()

32

The output tells us there are 32 unique seasons in the dataset. 

## **Q2**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

When calling this function on the Wteam attribute, it will give us the count of each value in descending order, so it will rank the teams by amount of wins. 

In [175]:
df['Wteam'].value_counts()

Wteam
1181    819
1242    804
1246    765
1314    761
1112    746
       ... 
1101     18
1446     14
1118      6
1289      6
1327      3
Name: count, Length: 364, dtype: int64

The output above indicates that team number 1181 has the most wins, but if we wanted to just directly get this output we could write this function:

In [176]:
df['Wteam'].value_counts().head(1)

Wteam
1181    819
Name: count, dtype: int64

Adding the `head(1)` function, which extracts only the first row, we know that we have extracted the team with the most wins. But this is only because the `value_counts()` function works in descending order!

# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [177]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [178]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [179]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [180]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [181]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [182]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [183]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [184]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [185]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [186]:
df.sort_values('Lscore').head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
100027    2008      66   1203      49   1387      20    H      0
89021     2006      44   1284      41   1343      21    A      0
49310     1997      66   1157      61   1204      21    H      0
85042     2005      66   1131      73   1216      22    H      0
103660    2009      26   1326      59   1359      22    H      0

## **Q3**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

##### $\underline{Answer}$

The first thing to do is make the 3 dataframes with the different ways of sorting. Here they are sorted in ascending order. This is done with the following: 

In [187]:
season = df.sort_values('Season') 
wteam = df.sort_values('Wteam')
wscore = df.sort_values('Wscore')

In [188]:
# Check if they look as expected 
print(season)
print(wteam)
print(wscore)

        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
159       1985      32   1339      46   1186      44    H      0
9         1985      25   1305      97   1424      89    H      0
10        1985      25   1307     103   1288      71    H      0
11        1985      25   1344      75   1438      71    N      0
12        1985      25   1374      91   1411      72    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145204    2016     129   1380      83   1106      63    N      0
145251    2016     130   1408      72   1222      69    N      0
145267    2016     131   1242      81   1452      71    N      0
145221    2016     130   1173      69   1350      54    N      0
145285    2016     132   1163      72   1272      58    N      0

[145289 rows x 8 columns]
        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
139440    2015     120   1101      83   1223      71    H      0
142816    2016      78   1101      75   1270      67    H      

Then we want to select rows from index 100 to 200 in each of the three dataframes, but only inlcude the columns for the attributes we are interested in; so Season, Wteam and Wscore

In Python and iloc, there is used 0-indexing, so to extract index 100 to 200, we will write 100:201 in the brackets, because 100 is included and 201 is excluded (so includes up to 200). If we had used loc, 100:200 would have been sufficient, as it includes both indices. 

In [189]:
selected1 = season.iloc[100:201, :][['Season', 'Wteam', 'Wscore']]
selected2 = wteam.iloc[100:201, :][['Season', 'Wteam', 'Wscore']]
selected3 = wscore.iloc[100:201, :][['Season', 'Wteam', 'Wscore']]

In [190]:
# Check if they look as expected 
print(selected1)
print(selected2)
print(selected3)

     Season  Wteam  Wscore
284    1985   1380      84
285    1985   1419      73
286    1985   1426      76
287    1985   1431      68
272    1985   1261      87
..      ...    ...     ...
380    1985   1449      68
381    1985   1450      62
382    1985   1461      73
383    1985   1113      81
368    1985   1248      45

[101 rows x 3 columns]
        Season  Wteam  Wscore
124488    2013   1102      86
130430    2014   1102      94
88409     2006   1102      64
86504     2005   1102      59
144295    2016   1102      76
...        ...    ...     ...
54503     1998   1102      80
114695    2011   1102      73
30862     1992   1102      60
32738     1993   1102      69
85718     2005   1102      52

[101 rows x 3 columns]
        Season  Wteam  Wscore
91997     2006   1143      43
75921     2003   1106      43
123683    2012   1450      43
9889      1987   1382      43
37148     1994   1391      43
...        ...    ...     ...
116389    2011   1197      45
46176     1996   1187      4

## **Q4**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

##### $\underline{Answer}$

The substets are selected1, selected2 and selected3. So we are interested in looking into the subset with the highest winning scores, so we will be using selected3 as this subsets is sorted by winning scores. Now we want to find the row in this subset with the maximum score. For this we can use the `.argmax()` function 

In [191]:
# Retrieve the index position for the highest winning score
hscore_index = selected3['Wscore'].values.argmax()
hscore_index

63

Now we can use the `.iloc` function to retrieve the values from row 63, which contains the highest winning score

In [192]:
hscore = selected3.iloc[hscore_index]
hscore

Season    2012
Wteam     1366
Wscore      45
Name: 120453, dtype: int64

From this output we can read that highest winning score was obtained by team 1366 and that it happened in season 2012. The score was 45. 

# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [193]:
df[df['Wscore'] > 150]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
5269     1986      75   1258     151   1109     107    H      0
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
16040    1989      40   1328     152   1331     122    H      0
16853    1989      68   1258     162   1109     144    A      0
17867    1989      92   1258     181   1109     150    H      0
19653    1990      30   1328     173   1109     101    H      0
19971    1990      38   1258     152   1109     137    A      0
20022    1990      40   1116     166   1109     101    H      0
22145    1990      97   1258     157   1362     115    H      0
23582    1991      26   1318     152   1258     123    N      0
24341    1991      47   1328     172   1258     112    H      0
24970    1991      68   1258     186   1109     140    H      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
40060    1995      32   1375     156   1341     114    H      0
52600    1998      33   1395     153   1410      87    H      0

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [194]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
12046    1988      40   1328     152   1147      84    H      0
12355    1988      52   1328     151   1173      99    N      0
25656    1991      84   1106     151   1212      97    H      0
28687    1992      54   1261     159   1319      86    H      0
35023    1993     112   1380     155   1341      91    A      0
52600    1998      33   1395     153   1410      87    H      0

## **Q5**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

##### $\underline{Answer}$

This can be done in two simple steps. First we create the dataframe that contains the score differences by subtracting losing scores from winning scores. We remember to use the abs() function to ensure positive values. Then we creathe filtered dataframe by using a Boolean statement; if the score difference of this observation is greater than the mean of all score differences, include them in the fultered dataframe. Else, exclude them. I have provided the code for this below

In [195]:
# 1. Creating the new dataframe ScoreDifference
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# 2. Create the filtered dataframe 
filtered_df = df[df['ScoreDifference'] > df['ScoreDifference'].mean()]

print(filtered_df)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
1

## **Q6**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

##### $\underline{Answer}$

To do this, we use the `.iloc()` and `.argmax()` function on the filtered dataframe which we just created. We want to extract values form the Season, Wteam and Lteam columns/attributes.

In [196]:
# Find game with highest score difference 
hscore_diff = filtered_df.iloc[filtered_df['ScoreDifference'].values.argmax()]
print(hscore_diff) # this actully gives us all desired information 

Season             1996
Daynum               48
Wteam              1409
Wscore              141
Lteam              1341
Lscore               50
Wloc                  H
Numot                 0
ScoreDifference      91
Name: 44653, dtype: object


For good measure, we can also call upon this function to get the season and relevnt teams involved in the game with the highest score difference.

In [197]:
# Find the season and two teams involved in the game 
hscore_diff.loc[['Season', 'Wteam', 'Lteam']]

Season    1996
Wteam     1409
Lteam     1341
Name: 44653, dtype: object

# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [198]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [199]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [200]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [201]:
df.values[0][0]

1985

## **Q7**: Group the DataFrame by season and find the average winning score for each season.

##### $\underline{Answer}$

To do this, we can use the same method as in codeline [64] above, but this time we will group by the Season attribute and not Wteam, so we will replace that in the soft parentheses. I have provided the code below. 

In [202]:
df.groupby('Season')['Wscore'].mean().head()

Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
Name: Wscore, dtype: float64

## **Q8**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

##### $\underline{Answer}$

Now we are interested in grouping by Wtean, and we will use the function `.max()` instead of `.mean()` to find the maximum value in a group/column. We don't use `.argmax()` because it doens't fit with the framework of the `.groupby()` function. I have provided the code below.  

In [203]:
df.groupby('Wteam')['Wscore'].max().head()

Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
Name: Wscore, dtype: int64

## **Q9**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

##### $\underline{Answer}$

Now we taking two categories/attributes into the `.groupby()` argument. The first thing we want to do is calculate the average score for every winning team in each season (avg), then we want to identify the team with the highest average winning score for every season (haws) and lastly we want to output the scores for the winning teams with the highest averages for each season (k). I have provided the code for this below. 

In [204]:
# Step 1 
avg = df.groupby(['Season','Wteam'])['Wscore'].mean()

# Step 2 
haws = avg.groupby('Season').idxmax()

# Step 3
k = avg.loc[haws]
k


Season  Wteam
1985    1328      92.800000
1986    1109      91.200000
1987    1380      95.875000
1988    1258     111.750000
1989    1258     117.315789
1990    1258     126.347826
1991    1380     112.312500
1992    1380      99.642857
1993    1380     101.875000
1994    1380     106.583333
1995    1206     102.833333
1996    1206     103.600000
1997    1254      94.142857
1998    1395     101.153846
1999    1317      95.750000
2000    1395      95.647059
2001    1395      99.315789
2002    1242      92.888889
2003    1395      88.222222
2004    1194      92.142857
2005    1311      91.333333
2006    1190      89.533333
2007    1440      99.400000
2008    1440      94.500000
2009    1377      96.200000
2010    1440      95.000000
2011    1449      90.130435
2012    1370      87.400000
2013    1377      86.500000
2014    1322      92.600000
2015    1322      90.733333
2016    1146      97.142857
Name: Wscore, dtype: float64

## **Q10**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

##### $\underline{Answer}$

We create a new dataframe still using the foundation of the `.groupby()` framework, which includes the `.count()` function, that very conveniently allows us to count the number of wins for each team in each season. I have provided the code for this below.

In [205]:
df.groupby(['Season', 'Wteam'])['Wscore'].count()

Season  Wteam
1985    1102      5
        1103      9
        1104     21
        1106     10
        1108     19
                 ..
2016    1460     20
        1461     12
        1462     27
        1463     21
        1464      9
Name: Wscore, Length: 10172, dtype: int64

## **Q11**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

##### $\underline{Answer}$

We can reuse the dataframe we made in codeline [93] for the first step. Then we want to identify the team with the most wins for every season (most_wins) and at last we want to output the team with the most wins for every season (h). I have provided the code below. 

In [206]:
# Step 1
win_count = df.groupby(['Season', 'Wteam'])['Wscore'].count()

# Step 2 
most_wins = win_count.groupby('Season').idxmax()

# Step 3
h = win_count[most_wins]
h

Season  Wteam
1985    1385     27
1986    1181     32
1987    1424     33
1988    1112     31
1989    1328     28
1990    1247     29
1991    1116     30
1992    1181     28
1993    1231     28
1994    1163     27
1995    1116     26
1996    1269     31
1997    1242     31
1998    1242     33
1999    1181     32
2000    1409     29
2001    1181     29
2002    1153     30
2003    1166     29
2004    1390     29
2005    1228     32
2006    1181     30
2007    1242     30
2008    1272     33
2009    1272     31
2010    1242     32
2011    1242     32
2012    1246     32
2013    1211     30
2014    1455     33
2015    1246     34
2016    1242     29
Name: Wscore, dtype: int64

## **Q12**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

##### $\underline{Answer}$

To approach this task, I will make two separate groupings of the dataframe, which gives us the average losing score for each team across seasons, and then the same for winning teams (losing and winning functions below). Then we want to compare and find if there are teams with higher losing than winning scores. For this I decided a Boolean statement would be nice, this allows us to loop over the values in the higher_losing_teamd dataframe, and then return all the values, where the dataframe has values. 

In [226]:
losing = df.groupby('Lteam')['Lscore'].mean()
winning = df.groupby('Wteam')['Wscore'].mean()

# combine into a single dataframe
score_comparison = pd.DataFrame({'avg_losing': losing,'avg_winning': winning})

# find where avg_losing > avg_winning 
higher_losing_teams = score_comparison[score_comparison['avg_losing'] > score_comparison['avg_winning']]

if not higher_losing_teams.empty: 
    print(higher_losing_teams)
else: 
    print('All teams won more than they lost !')

All teams won more than they lost !


As can be seen, the dataframe is empty, which means that no scores were losing scores were found to be higher than winning scores for any of the teams. 

# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [227]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
HighScoringGame      No
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
HighScoringGame      No
Name: 1, dtype: object


## **Q13**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

##### $\underline{Answer}$

The easiest approach we can do here is to create a list in which we will store all the games, but now associate them with either a Yes or No depending on they scored greater than 100 points or not. So, I start by cretaing an empty list (high_score). Then, I use the `.iterrows()` function for dataframes, to loop over the rows in the dataframe; I make a Boolean condition that takes value of the Wscore at that specific index, and if if the numerical value is above 100, I append the text "Yes" to the high_score list. If it is below 100, I append the the text "No" to the high_score list. After iterating through the entire dataframe and accumulating the list, I then add the list to the dataframe as a new column/attribute. At last, we can display the results of our work by inputing the Wscore and GameWithHighScore columns/attributes as our dataframe arguments.

In [238]:
# Step 1 
high_score = []

# Step 2
for index, row in df.iterrows():
    if row['Wscore'] > 100:
        high_score.append('Yes')
    else:
        high_score.append('No')

# Step 3
df['GameWithHighScore'] = high_score
df[['Wscore', 'GameWithHighScore']]

Wscore GameWithHighScore
0           81                No
1           77                No
2           63                No
3           70                No
4           86                No
...        ...               ...
145284      70                No
145285      72                No
145286      82                No
145287      66                No
145288      87                No

[145289 rows x 2 columns]

This is not very interesting, where are all the yes'es? We can find this out by filtering the dataframe for games where they are "Yes" and "No" respectively. We can use the `.loc` function for this. 

In [239]:
df.loc[df['GameWithHighScore'] == 'Yes', ['Wscore', 'GameWithHighScore']]

Wscore GameWithHighScore
10         103               Yes
25         106               Yes
35         109               Yes
42         104               Yes
50         102               Yes
...        ...               ...
144999     108               Yes
145135     101               Yes
145148     108               Yes
145218     104               Yes
145245     104               Yes

[4880 rows x 2 columns]

In [240]:
df.loc[df['GameWithHighScore'] == 'No', ['Wscore', 'GameWithHighScore']]

Wscore GameWithHighScore
0           81                No
1           77                No
2           63                No
3           70                No
4           86                No
...        ...               ...
145284      70                No
145285      72                No
145286      82                No
145287      66                No
145288      87                No

[140409 rows x 2 columns]

## **Q14**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

##### $\underline{Answer}$

We are now using a dictionary which is convenient, because we can store the teams as keys and the number of games they've played as values. Then we iterate as before, but now we append to the ngames dictionary with the number 1 for every occurence of that team. We do this for both losing and winning teams. It is not so easy to just add a dictionary to a dataframe as a column (like with lists) so i will do something different. Now, we don't care if the team is winning or losing, we just want to know the team number and the number of games they've playes. So, I will convert the dictionary we've created to a new dataframe, where we assign two columns Team and PlayedGames. To this we essentially force the dictionaty to be a list by converting it into a list of key-value pairs (very convenient !). I think it makes most sense to present the dataframe in descending order, so seeing which teams played most games at the top and least at the bottom. This can be done by using the `.sort_values()` function, but since it sorts by acsending order by defalult, we need to add `ascending=False` to give us the desired sorting method. 

In [251]:
# Step 1 
ngames = {}

for index, row in df.iterrows(): 
    if row['Wteam'] in ngames: 
        ngames[row['Wteam']] += 1
    else: 
        ngames[row['Wteam']] = 1

    if row['Lteam'] in ngames: 
        ngames[row['Lteam']] += 1
    else: 
        ngames[row['Lteam']] = 1

# Step 2 
ngames_df = pd.DataFrame(list(ngames.items()), columns=['Team', 'PlayedGames'])

# Step 3
most_games = ngames_df.sort_values(by='PlayedGames', ascending=False)
most_games
        

Team  PlayedGames
106  1181         1013
98   1314         1010
136  1246          999
61   1257          994
168  1393          993
..    ...          ...
363  1230           64
329  1289           48
249  1446           48
285  1118           27
278  1327           22

[364 rows x 2 columns]

## **Q15**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

##### $\underline{Answer}$

What we will do is create a dictionary which keys are Season and the corresponding value will be the highest ScoreDifference (hsd). Then we loop over the dataframe saying that if the Season already exists in hsd, we compare the current ScoreDifference with the value already there. If it happens that this ScoreDifference is greater than the stored one, we update highest ScoreDifference for that Season in the Dictionary. If the Season is not yet in the dictionary, we add it to it and the corresponding ScoreDifference. 

After the iteration is over, we have a dictionary that stores all the seasons (1985-2016) and the game with the highest ScoreDifference. 

In [252]:
df.columns.tolist()

['Season',
 'Daynum',
 'Wteam',
 'Wscore',
 'Lteam',
 'Lscore',
 'Wloc',
 'Numot',
 'ScoreDifference',
 'HighScoringGame',
 'GameWithHighScore',
 'GamesInTotal']

In [259]:
hsd = {}

for index, row in df.iterrows():
   if row['Season'] in hsd: 
     if row['ScoreDifference'] > hsd[row['Season']]:
        hsd[row['Season']] = row['ScoreDifference']

   else: 
      hsd[row['Season']] = row['ScoreDifference']

hsd

{1985: 60,
 1986: 84,
 1987: 73,
 1988: 68,
 1989: 70,
 1990: 76,
 1991: 68,
 1992: 82,
 1993: 81,
 1994: 69,
 1995: 74,
 1996: 91,
 1997: 80,
 1998: 66,
 1999: 75,
 2000: 75,
 2001: 75,
 2002: 59,
 2003: 72,
 2004: 70,
 2005: 64,
 2006: 72,
 2007: 74,
 2008: 70,
 2009: 68,
 2010: 75,
 2011: 69,
 2012: 62,
 2013: 58,
 2014: 74,
 2015: 69,
 2016: 73}

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [260]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

## **Q16**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

##### $\underline{Answer}$

Okay, now no more looping. To approach this task I will use the `value_counts()` function as instructed, as well as the `.add()` function. 

In [263]:
won = df['Wteam'].value_counts()
lost = df['Lteam'].value_counts()

tn_games = won.add(lost)
tn_games

1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: count, Length: 364, dtype: int64

## **Q17**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

##### $\underline{Answer}$

Since we already created the ScoreDifference column in a task above (Q5), it would be superflous to do it again. I start by making a function called hsd which is grouped by the Seasons and then finds the index where the highest score difference occurs. Then we locate and extract the rows we want by using the `.loc` function. 

In [268]:
hsd = df.groupby('Season')['ScoreDifference'].idxmax()
df.loc[hsd, ['Season', 'ScoreDifference']]

Season  ScoreDifference
236       1985               60
4731      1986               84
8240      1987               73
12046     1988               68
16677     1989               70
19502     1990               76
25161     1991               68
27997     1992               82
33858     1993               81
36404     1994               69
39858     1995               74
44653     1996               91
49033     1997               80
52600     1998               66
57618     1999               75
60616     2000               75
65999     2001               75
70737     2002               59
74791     2003               72
79181     2004               70
83372     2005               64
89542     2006               72
94507     2007               74
97737     2008               70
103723    2009               68
108405    2010               75
114015    2011               69
118699    2012               62
124057    2013               58
131373    2014               74
134590    2015               69
140040    2016               73

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [269]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [270]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [271]:
type(df['Wscore'])

pandas.core.series.Series

In [272]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [273]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame GameWithHighScore  GamesInTotal  
0              No                No          1960  
1              No                No          1761  
2              No                No          1344

Here's an equivalent using iloc

In [274]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame GameWithHighScore  GamesInTotal  
0              No                No          1960  
1              No                No          1761  
2              No                No          1344

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [275]:
df.isnull().sum()

Season               0
Daynum               0
Wteam                0
Wscore               0
Lteam                0
Lscore               0
Wloc                 0
Numot                0
ScoreDifference      0
HighScoringGame      0
GameWithHighScore    0
GamesInTotal         0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y